Implement sdp module

alexandremgo opened the pull request:
Working on issue #200 

First implementation of the parser and the SessionDescription struct.

Can you check my code before I continue the implementation to be sure I'm working on goods foundations ? Thanks!

gavv commented:
Looks pretty good.

A few notes. Don't know what is WIP and what is final here, so just listing everything, except grammar.

* Yes, I suggest to hide GUID building logic inside SessionDescription.

* I suggest to hide all public fields of SessionDescription and provide only public setters and getters, like we do it in address::IoURI and address::EndpointURI. How fields are allocated and stored will be an implementation detail.

* IAllocator reference will be passed to SessionDescription constructor.

* If parse_sdp() will need to allocate temporary Arrays, IAllocator reference should be also passed to parse_sdp(). Allocator will be a long-living object tied to a session of for entire roc_context. Its specific type should not be hardcoded.

* I think we can avoid temporary arrays in parse_sdp(). Instead of copying string from [start_p; p] to temporary array in those actions, we can just store current start_p and p pointers and later pass these pointers to SessionDescription::set_guid(). (These pointers point to a substring of the input string). set_guid() will perform all necessary allocations and copying. This way we will reduce number of allocations and copies because set_guid() can do only a single allocation. This way we can also make all allocations and copying the implementation detail of SessionDescription, making the parser code cleaner.

* We should always check return value of resize(), copy_str(), and append_str(). In places when it should never fail, we should still check return value and panic if it failed.

* When parsing fails, we should clear() SessionDescription before returning false to ensure that we will not return partially filled description (to decrease probability of errors in upper code).

* Regarding AddrType enum. In SocketAddr, we have: `set_host_port_ipv4()`, `set_host_port_ipv6()`, `set_miface_ipv4()`, `set_miface_ipv6()`, and `int version()` (which returns `4` or `6`). In roc_sdp we're adding address type and then a switch that calls the right method according to the type. Maybe move AddrType enum to roc_address and replace `set_host_port_ipv4(...)` with `set_host_port(type, ...)`, and so on? We can also change `version()` to return enum instead of int. In other words, move the switch into SocketAddr. I think this will be useful to another SocketAddr users too and make the code cleaner. This can be done in a separate PR.

* Regarding to the grammar itself, I guess it's still WIP, so it's too eraly to review it?

alexandremgo commented:
Thank you for the detailed remarks :) 

>     * Regarding to the grammar itself, I guess it's still WIP, so it's too eraly to review it?

Yes it's too early ;)

gavv commented:
While working on #258, I've added two new classes to roc_core:

* `StringBuffer`, for storing strings (7c3b918bbb4d2f5029fd8443bcf8e5ec378c9674). It's a wrapper for `Array<char>` that guarantees that the string is always zero-terminated.

* `StringBuilder`, for building strings, no matter how they're stored (b0e4567a3931d4a8fa130a5effe1c94e87be3bff). It's a replacement for `string_utils.h` (copy_str, append_str, etc).

I suggest to use `StringBuffer` for storing session description fields and use `StringBuilder` when formatting session description into a string. See how they're used in `address::EndpointURI`.

I'd like to remove `string_utils.h`, but kept it to so as not to break your PR. Let me know when you don't need it anymore.

gavv commented:
(And also let me know if there are any problems with the suggested approach)

alexandremgo commented:
> (And also let me know if there are any problems with the suggested approach)

I think the suggested approach is good. I just had some difficulty to find enough time to work on the PR recently. But I'm working on it ;)

alexandremgo commented:
I think we should close this PR and I'll re-open one with a clean branch (the rebase is difficult with all the changes that occurred) once I've added your suggestion ?

gavv commented:
As you wish. You can just hard-reset the branch to any other commit and then force-push it, but opening a new PR is OK too, why not.

alexandremgo commented:
How can we handle the TTL from an IPv4 multicast address ? It can appear in the  `connection-address` of the connection field (https://tools.ietf.org/html/rfc4566#page-14). For now my grammar concerning unicast and multicast address is broad (because according to the RFC 4566 an address can be something else than an IPv4/IPv6).

gavv commented:
IIRC, TTL is deprecated in SDP. The grammar should allow it, but the code should ignore it. SDP allows TTL only for IPv4. So our grammar can have a branch for "IN IP4" which has optional "/number" field after host.

alexandremgo commented:
Hey! Is the use of `core::List` in `SessionDescription` for the `media_descriptions` and in `MediaDescription` for the `connection_fields` good ?
For now the data of each media is saved as a string. Should we store the data in different variables ?

gavv commented:
Hi, sorry for delay (again)!

1. I think using core::List for MediaDescription is okay. MediaDescription will likely contain dynamic structures like rtp::FormatMap, so preferably it should not be copyable, and thus we should allocate it on heap. And core::List is well suited for such objects.

2. As for the ConnectionField, we could also use core::List, but since this struct is small and fixed-size, we can instead make it copyable and use core::Array<ConnectionField, 1> to avoid extra allocations. (So we'll have zero additional allocations for `c=` in most cases). Note that address::SocketAddr is already copyable.

3. Also, we could rename ConnectionField to ConnectionData, as it is called in the RFC.

4. As for the MediaDescription fields. Yes, we need to store (and access) `m=` fields separately. E.g. we'll need to get the port number, profile, and payload id. I guess we also need a enum for profiles (the proto sub-field of `m=`), for now with a single supported value for `RTP/AVP` profile.

5. In addition, we'll also need to parse various attributes and add them to MediaDescription. For example, we should use `a=rtpmap` to build rtp::FormatMap and `a=a=fec-source-flow` / `a=a=fec-repair-flow` to build fec::CodecConfig. We could either store this attributes in some generic key-value form in MediaDescription and then convert them to rtp::FormatMap and fec::CodecConfig somewhere else; or we can directly store like rtp::FormatMap and fec::CodecConfig in MediaDescription and fill them when is MediaDescription setters. I think the second approach is simpler and thus preferable, so I suggest to try it first and see how it will look like.

6. I think we shouldn't duplicate MediaDescription setters in SessionDescription. I mean the add_connection_to_last_media() method. Instead, I think the parser should either fully build MediaDescription and then add it to SessionDescription, or ask SessionDescription to create a new MediaDescription, get a reference to it (e.g. using last_media_description), and use its setters directly. Not sure what approach is better.

gavv commented:
Let me know when the PR will be (or is already?) ready for a more detailed review.

gavv commented:
At a quick glance, the current code looks pretty nice :)

alexandremgo commented:
Hey! No pb, it has also been difficult for me recently. Thanks for your comments!

I (almost ?) finished the main fields (v, o, c, m) and I'm going to add the different attributes soon.

Concerning your last comments:

2. I used core::Array<ConnectionData, 1> as advised for the ConnectionData in the MediaDescription.
6. I implemented the second approach, using last_media_description.

gavv commented on src/modules/roc_sdp/connection_data.cpp:
```diff
@@ -0,0 +1,39 @@
+    char addr[address::SocketAddr::MaxStrLen];
+    core::StringBuilder b(addr, sizeof(addr));
+
+    if (!b.append_str_range(str, str + str_len))
+        return false;
```
Nitpick: we always use {} even for single-statement blocks.

gavv commented on src/modules/roc_sdp/connection_data.cpp:
```diff
@@ -0,0 +1,39 @@
+    if (!b.append_str_range(str, str + str_len))
+        return false;
+
+    roc_log(LogInfo, "Connection Field address: %s", &addr[0]);
```
Nitpick: log messages usually include component name and are in the lower case, e.g.:

    "sdp: connection field address: %s"

gavv commented on src/modules/roc_sdp/media_description.cpp:
```diff
@@ -0,0 +1,109 @@
+}
+
+bool MediaDescription::add_fmt(const char* str, size_t str_len) {
+    core::StringBuffer<> fmt(allocator_);
+    if (!fmt.set_buf(str, str_len) || fmt.is_empty()) {
+        return false;
+    }
+
+    if (!fmts_.push_back(fmt.c_str())) {
+        return false;
+    }
```
Maybe add `StringList.push_back_range(begin, end)` to get rid of an extra copy?

gavv commented on src/modules/roc_sdp/media_description.cpp:
```diff
@@ -0,0 +1,109 @@
+        return false;
+    }
+
+    if (connection_data_.size() >= connection_data_.max_size()) {
+        if (!connection_data_.grow(connection_data_.size() + 1)) {
```
No need for extra check `if (connection_data_.size() >= connection_data_.max_size())`, grow() already performs it. Also, maybe use grow_exp(), since we're iteratively appending here?

gavv commented:
> I (almost ?) finished the main fields (v, o, c, m) and I'm going to add the different attributes soon.

Let me know if you want to move the rest attributes to a separate PR, in this case I can perform the final review of this one. Or we can merge all at once, up to you.

alexandremgo commented:
> Let me know if you want to move the rest attributes to a separate PR, in this case I can perform the final review of this one. Or we can merge all at once, up to you.

Yes, let's put the other attributes to a separate PR. I think it's ready for a more in depth review ;)

gavv commented on src/modules/roc_sdp/parser.rl:
```diff
@@ -0,0 +1,283 @@
+                return false;
+            }
+
+            if (!result.last_media_description()->set_nb_ports((int)nb_ports)) {
```
We always use `last_media_description()` without checking whether it returned NULL. Is it guaranteed by our grammar that we can call it only after at least one `create_media_description()` call? I guess yes, just to be sure.

alexandremgo replied:
Yes it's guaranteed by our grammar: add_media is always called first, and if the media could not be created, we return false and stop the parsing.

gavv replied:
Thanks, great.

gavv commented on src/modules/roc_sdp/parser.rl:
```diff
@@ -0,0 +1,283 @@
+        # typically an RTP payload type for audio and video media
+        media_fmt = token >start_token %add_media_fmt;
+
+        media_description = (media_type ' ' media_port ('/' media_nb_ports)? ' ' media_proto (' ' media_fmt)+)
```
Does SDP allows only a single space as delimiter?

alexandremgo replied:
According to the [RFC 5234](https://tools.ietf.org/html/rfc5234#page-14) SP represents a single space. I added a definition in the parser.

gavv commented:
@alexandremgo https://github.com/roc-project/roc/issues/200#issuecomment-612383960

alexandremgo commented on src/modules/roc_sdp/session_description.cpp:
```diff
@@ -0,0 +1,103 @@
+    return guid_.c_str();
+}
+
+bool SessionDescription::set_guid(const char* start_p_origin_username,
+                                  const char* end_p_origin_sess_id,
+                                  const char* start_p_origin_nettype,
+                                  const char* end_p_origin_addr) {
```
From the RFC:
"In general, the "o=" field serves as a globally unique identifier for this version of this session description, and the subfields excepting the version taken together identify the session irrespective of any modifications". 

So I remove the  `sess-version` part of the origin field, that is between the `username` and the `sess-id`, to create the GUID.

But I can add it, and the GUID will be just start_p_origin_username and end_p_origin_addr (so the entire origin field).

What do you think ?

From the LogInfo i put in this function, addrtype seems to be taken into account in the GUID.

gavv replied:
Excluding `sess-version` from GUID is correct.

What I'm asking, why do we, for example, add the entire string `<username> <sess-id>` (`append_str_range(start_p_origin_username, end_p_origin_sess_id)`) instead of adding `<username>` and `<sess-id>` separately, excluding separator (space) between them?

The current approach assumes that the separator is always the same, e.g. is always exactly one space, and it can not be two spaces, or other white-space character, like tabulation. Is this true and required by RFC? If yes, the code is good, otherwise, two logically equal descriptions, which differ only by some delimiters, may produce different GUIDs, which is bad.

gavv replied:
> From the LogInfo i put in this function, addrtype seems to be taken into account in the GUID.

Indeed, my bad.

gavv replied:
Yeah, and even if RFC requires exactly one space, does it make sense to be ready for multiple spaces as delimiters anyway? I can imagine that some implementation may produce extra spaces, though I'm not aware of such. Usually it's a good idea to be generous when parsing and allow "almost" correct input while it remains unambiguous. Would such behavior break other things?

alexandremgo replied:
Yeah you're right, it's better to be ready for multiple spaces or a tab as delimiters.

gavv replied:
Added this to https://github.com/roc-project/roc/issues/200#issuecomment-612383960

alexandremgo commented on src/modules/roc_sdp/session_description.cpp:
```diff
@@ -0,0 +1,103 @@
+        return false;
+    }
+
+    b.append_char(' ');
```
Yes we can! What would you use ?
In the RFC they don't really talk about the guid (or I missed it ?), except in the quote from the previous comment.

And for the parameters of set_guid I would have an start_p and end_p for each variable if we do a non-space delimiter.

gavv replied:
To produce GUIDs unambiguously, we should use a delimiter that is prohibited from use in GUID fields. So we can use any character that can't be used in `o=` sub-fields (username, sess-id, etc). Space is good, but it would be more convenient if GUIDs didn't contain spaces. Are there any other characters prohibited by RFC in `o=` sub-fields? Preferably, some printable ASCII character.

alexandremgo replied:
In the origin field, a username is a non-ws-string, which is defined by:  1*(VCHAR/%x80-FF) = string of visible characters.

The [RFC 4566](https://tools.ietf.org/html/rfc4566#section-9) comments on the username field: "pretty wide definition, but doesn't include space".

In [RFC 5234](https://tools.ietf.org/html/rfc5234), they define VCHAR = %x21-7E.

So except a space, I don't think there is any ASCII character that is available :/

gavv replied:
Thanks, I see. Then let's keep using space as delimiter.

gavv commented on src/modules/roc_sdp/parser.rl:
```diff
@@ -0,0 +1,310 @@
+    write data;
+}%%
+
+bool parse_sdp_imp(const char* str, SessionDescription& result);
```
We should either declare it static or, better, define it in an anonymous namespace.

gavv commented:
Resolved most conversations, ping me when this will be ready for review.

alexandremgo commented:
> Resolved most conversations, ping me when this will be ready for review.

@gavv it's ready for review ;)

gavv commented on src/modules/roc_sdp/parser.rl:
```diff
@@ -0,0 +1,283 @@
+        media_fields = ('\n' media_field ('\n' media_connection_field)*)+;
+
+        sdp_description = 'v='i version
+        '\n' 'o='i origin
```
Added this to https://github.com/roc-project/roc/issues/200#issuecomment-612383960

gavv commented on src/modules/roc_sdp/parser.rl:
```diff
@@ -0,0 +1,310 @@
+
+            media_fields = (CRLF media_field (CRLF media_connection_field)*)*;
+
+            sdp_description = 'v='i version
+            CRLF 'o='i origin
+            (CRLF 'c='i session_connection_data)?
+            media_fields;
```
```
   An SDP session description consists of a number of lines of text of
   the form:

      <type>=<value>

   where <type> MUST be exactly one case-significant character
```

Why do we use `i` here and in other places, e.g. `'v='i`?

alexandremgo replied:
It was to make them case-insensitive, but it should actually be case-sensitive. I'll remove them.

gavv commented on src/modules/roc_sdp/parser.rl:
```diff
@@ -0,0 +1,310 @@
+            
+            media_field = 'm='i media_description;
+
+            # In media-level: c=<nettype> <addrtype> <connection-address>
+            media_connection_nettype = "IN";
+            
+            # Either IPv6 or IPv4/TTL
+            media_connection_address = non_ws_string >start_token 
+                %add_media_connection_data;
+
+            media_connection_with_addrtype =  
+                ( "IP4" %{ cur_addrtype = address::Family_IPv4; } 
+                    SP media_connection_address '/' digit+
+                | "IP6" %{ cur_addrtype = address::Family_IPv6; } 
+                    SP media_connection_address
+                );
+
+            media_connection_data = media_connection_nettype SP 
+                media_connection_with_addrtype;
+
+            media_connection_field = 'c='i media_connection_data;
```
Could we avoid duplicating grammar for media- and session-level connection data?

alexandremgo commented on src/tests/roc_sdp/test_sdp_parse.cpp:
```diff
@@ -0,0 +1,47 @@
+
+    MediaDescription* last_media = session_description.last_media_description().get();
+
+    CHECK_EQUAL(6789, last_media->port());
+    CHECK_EQUAL(MediaType_Audio, last_media->type());
+    CHECK_EQUAL(MediaTransport_RTP_AVP, last_media->proto());
+    CHECK_EQUAL(10, last_media->default_payload_id());
```
What would be best to retrieve the MediaDescription from the SessionDescription: should we return a reference to the core::List<MediaDescription> ? Or a getter that take a number/position in the list and return a reference to the related MediaDescription ?

Same to retrieve the ConnectionData from a MediaDescription: should we return a reference to the core::Array<ConnectionData> or a getter that take an id and return a reference to the related ConnectionData ?

gavv replied:
Regarding ConnectionData: I guess we should encapsulate how it is stored (array or something), since there are multiple forms how the same logical value can be represented (multiple `c=` lines or a single `c=` line with number-of-addresses field, see https://github.com/roc-project/roc/issues/200#issuecomment-612383960). So maybe in this case we should somehow abstract connection data by getters to handle both possible underlying representations transparently.

Regarding MediaDescription. A getter for retrieving MediaDescription by its position (index) would be inefficient (it'd require linear search), so we either can return a const reference to core::List or replicate its nextof() getter. TBH I don't see a big difference at this point. If you don't see problems with returning List, maybe this way is better because it requires less code.

alexandremgo commented:
Hey @gavv ! I tried to rebase to finish this pull request (or at least to have something that has no conflict with the current develop branch), but there were too many changes since the last time i touched this code. The rebase is hell, and did not work. Do you have any tips to have a clean branch pls ?

gavv commented:
There are two good strategies: 1) hard reset to develop head and cherry pick your commits; 2) run rebase and then just --skip all conflicting commits except your own.

Sorry for conflicts, I'll try to avoid frequent rebases in future. Instead of rebasing each time I need to deliver a fix to master, I'll cherry pick the fix and will rebase develop on master only when doing a release, or something like this.

gavv commented:
@alexandremgo BTW if you want, you can extract changes in roc_core into separate PR and we'll merge it quickly to avoid accumulating more conflicts.

gavv commented:
And sorry for conflicts again, I already forgot that StringList was modified in this PR :)

alexandremgo commented:
Hey @gavv ! I made the changes mentioned in your comments, except the abstraction to get the ConnectionData. I don't see how it can be done easily. That part could need another PR.

I don't see the usual travis tests on my last commits. I rebased correctly (i think) with the develop branch. Should i do something else to activate the travis tests ?

gavv commented:
Great! I'll review it soon.

Travis builds sometimes fail to link to PR. In this case you can find them manually here: https://travis-ci.org/github/roc-streaming/roc-toolkit/pull_requests
